### ISSUE TITLE
A System.registerDynamic dependency of a System.register module is not executed

### ISSUE THREAD

**Issue by davdiv** [systemjs/systemjs#652]
Hello,

Thank you for all the work you are doing on the systemjs loader.

I am opening this issue because I think there is a bug in System.js. I have a `main.js` module, declared with `System.register` (it could be the output of a transpilation done with babel):

```js
console.log("main.js: begin");
System.register(['cjsDependency.js'], function (_export) {
  console.log("main.js: inside System.register callback function");

  var cjsDependency;
  return {
    setters: [function (_cjsDependency) {
      cjsDependency = _cjsDependency['default'];
      console.log("main.js: inside cjsDependency setter function, typeof cjsDependency = " + (typeof cjsDependency));
    }],
    execute: function () {
      console.log("main.js: inside execute function, typeof cjsDependency = " + (typeof cjsDependency));
      cjsDependency(); // cjsDependency should be a function!
      _export('default', cjsDependency);
    }
  };
});
console.log("main.js: end");

```

As you can see, this module has a dependency on the `cjsDependency.js` module, which is declared with `System.registerDynamic`:

```js
console.log("cjsDependency.js: begin");
// System.registerDynamic is documented here:
// https://github.com/systemjs/systemjs/blob/master/docs/system-api.md#systemregisterdynamicname--deps-executingrequire-declare
System.registerDynamic([/* Common JS module with no dependency */], true, function(require, exports, module) {
  console.log("cjsDependency.js: inside registerDynamic callback function");
  var myFunction = function() {
    console.log("cjsDependency.js: inside exported function");
  };
  module.exports = myFunction;
});
console.log("cjsDependency.js: end");
```

When I load `main.js` with `System.import`, I would expect `cjsDependency.js` to be fully executed before the `execute` function of `main.js` is called. But it is not the case: the console never displays `inside registerDynamic callback function` and there is an exception happening at the line marked with the comment `// cjsDependency should be a function!`.
Am I doing something wrong?
You can have a look at the full source code here: https://github.com/davdiv/systemjsBug
You can reproduce the issue at the following URL: http://davdiv.github.io/systemjsBug/

Thank you in advance for your advice!

**Comment by guybedford** [comment]
Note that by adding `console.log` before the `System.register` and `System.registerDynamic` statement that would interfere with the module format detection. The way this detection works is explained in https://github.com/systemjs/systemjs/blob/master/docs/module-formats.md#module-format-detection.

**Comment by davdiv** [comment]
I have just removed them and the result is the same.

**Comment by davdiv** [comment]
I investigated the issue. It looks like removing the check for the `declarative` boolean [here](https://github.com/systemjs/systemjs/blob/master/dist/system-register-only.src.js#L1298) solves the problem (it allows not to loose the entry by setting the `anonRegister` variable):

```js
function doRegister(loader, name, register) {
    calledRegister = true;

    // named register
    if (name) {
      var ext = loader.defaultJSExtensions && name.split('/').pop().split('.').pop();
      name = (loader.normalizeSync || loader.normalize).call(loader, name);
      if (ext && name.substr(name.length - ext.length - 1, ext.length + 1) != '.' + ext)
        name = name.substr(0, name.lastIndexOf('.'));
      register.name = name;
      if (!(name in loader.defined))
        loader.defined[name] = register; 
    }
    // anonymous register
    else /* if (register.declarative) */ { // Removing this check solves the issue
      if (anonRegister)
        throw new TypeError('Invalid anonymous System.register module load. If loading a single module, ensure anonymous System.register is loaded via System.import. If loading a bundle, ensure all the System.register calls are named.');
      anonRegister = register;
    }
  }
```

**Comment by guybedford** [comment]
Released in 0.18.8.

**Comment by davdiv** [comment]
@guybedford Thank you!
